### DISCUSSION TITLE
WebAssembly and gRPC-Web

### DISCUSSION THREAD

**Discussion by t-sorger** [unoplatform/uno#8994] [Q&A] [answered]
How do I use gRPC-Web in my Wasm project with Uno Platform?

I am trying to setup a client-server application which uses gRPC as communication framework. 
My target platforms are Windows, Android and WebAssembly. For Windows and Android I can use gRPC in a console application and it works. But for Wasm I haven't found any good example/reference on how to use gRPC-Web in my Uno Application/Wasm project.

Any help/demo would be very appreciated.

**Comment by jeromelaban** [comment, marked as answer]
We do not yet have a sample for this, but it is likely that most of [the content provided by this article](https://azure.github.io/AppService/2021/03/15/How-to-use-gRPC-Web-with-Blazor-WebAssembly-on-App-Service.html) will apply to Uno as well.

**Reply by t-sorger** [reply]
Thank you for your answer @jeromelaban. I created a [sample application]( https://github.com/t-sorger/UnoGrpcWebAssemblyDemo) with [the article you have mentioned]( https://azure.github.io/AppService/2021/03/15/How-to-use-gRPC-Web-with-Blazor-WebAssembly-on-App-Service.html) and added another counter service as defined in [this gRPC-dotnet example]( https://github.com/grpc/grpc-dotnet/tree/master/examples/Counter).

After everything was working, I added an Uno Platform WebAssembly App which should replace the BlazorApp client in the future.

Currently I am stuck at how to start the Uno WebAssembly App correctly. 
The Uno Platform uses the following ```Start```-method:
```
Microsoft.UI.Xaml.Application.Start(_ => _app = new App());
```
The BlazorApp on the other hand uses a ```WebAssemblyHostBuilder``` in which you can add services with an according ```GrpcChannel``` and then start the application by calling the ```RunAsync```-method:
```
var builder = WebAssemblyHostBuilder.CreateDefault(args);

builder.Services.AddSingleton(services => new Counter.CounterClient(_channel));

await builder.Build().RunAsync();
```
I haven’t found the correct way of doing this in the Uno Platform WebAssembly App yet. By that, I haven’t figured out on how to start the application correctly.

Again, any help or advise from your side would be very appreciated.

**Reply by jeromelaban** [reply]
By default, Uno Applications do not use the `HostBuilder`, so you don't need to use this particular piece. You can use one if you want by using the `GenericHostBuilder`, but this is a simple DI (Dependency Injection) container. You only need to create your `CounterClient` class directly and use it in your project.

If you want a similar DI and IoC experience, you can also use our [`Uno.Extensions` template](https://platform.uno/docs/articles/external/uno.extensions/doc/Overview/ExtensionsOverviewAndGettingStarted.html), which uses the `GenericHost` to allow you to have a similar experience to Blazor or ASP.NET.

**Reply by t-sorger** [reply]
Thank you for this information @jeromelaban.
With the help of @carldebilly my [repository "UnoGrpcWebAssemblyDemo"](https://github.com/t-sorger/UnoGrpcWebAssemblyDemo) is finally functioning.

**Reply by jeromelaban** [reply]
Thanks. [For reference](https://stackoverflow.com/a/72581301/26346), the `WebAssemblyHostBuilder` cannot be used in Uno applications, as it relies on private javascript exports specific to Blazor.

**Comment by jpgarza93** [comment]
It would be very useful if this could be documented with a bit more info on the key components, things to keep an eye for. 
Maybe expand on the ReadMe? 

I am trying to get my GRPC client working on WASM. But my GRPC server is a C++ console application.

Thanks in advance.
